File: app/src/main/java/openfoodfacts/github/scrachx/openfood/utils/LocaleHelper.java
package openfoodfacts.github.scrachx.openfood.utils;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.os.Build;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import openfoodfacts.github.scrachx.openfood.views.OFFApplication;
import org.apache.commons.text.WordUtils;

import java.util.*;

/**
 * This class is used to change your application locale and persist this change for the next time
 * that your app is going to be used.
 * <p/>
 * You can also change the locale of your application on the fly by using the setLocale method.
 * <p/>
 * Created by gunhansancar on 07/10/15.
 */
public class LocaleHelper {
    public static int find(List<LanguageData> availableLanguageForImage, String language) {
        if (language != null) {
            for (int i = 0; i < availableLanguageForImage.size(); i++) {
                if (language.equals(availableLanguageForImage.get(i).code)) {
                    return i;
                }
            }
        }
        return -1;
    }

    public static class LanguageData implements Comparable<LanguageData> {
        private final String code;
        private final String name;
        private final boolean supported;

        LanguageData(String code, String name, boolean supported) {
            this.code = code;
            this.name = name;
            this.supported = supported;
        }

        public boolean isSupported() {
            return supported;
        }

        public String getCode() {
            return code;
        }

        public String getName() {
            return name;
        }

        @Override
        public String toString() {
            return name+" ["+code+"]";
        }

        @Override
        public boolean equals(Object obj) {
            if (obj == null) {
                return false;
            }
            if (LanguageData.class.equals(obj.getClass())) {
                return code.equals(((LanguageData) obj).code);
            }
            return false;
        }

        @Override
        public int hashCode() {
            return code.hashCode();
        }

        @Override
        public int compareTo(@NonNull LanguageData o) {
            return name.compareTo(o.name);
        }
    }

    private LocaleHelper() {
        //Helper class
    }

    private static final String SELECTED_LANGUAGE = "Locale.Helper.Selected.Language";
    public static final String USER_COUNTRY_PREFERENCE_KEY = "user_country";

    public static Context onCreate(Context context) {
        String lang = getLanguageInPreferences(context, Locale.getDefault().getLanguage());
        return setLocale(context, lang);
    }

    public static Context onCreate(Context context, String defaultLanguage) {
        String lang = getLanguageInPreferences(context, defaultLanguage);
        return setLocale(context, lang);
    }

    public static List<LanguageData> getLanguageData(Collection<String> codes, boolean supported) {
        List<LanguageData> res = new ArrayList<>();
        if (codes != null) {
            for (String code : codes) {
                final LanguageData languageData = getLanguageData(code, supported);
                if (languageData != null) {
                    res.add(languageData);
                }
            }
        }
        Collections.sort(res);
        return res;
    }

    public static LanguageData getLanguageData(String code, boolean supported) {
        Locale locale = getLocale(code);
        if (locale == null) {
            return null;
        }
        return new LanguageData(locale.getLanguage(), WordUtils.capitalize(locale.getDisplayName(locale)), supported);
    }

    public static Locale getLocale() {
        return getLocale(OFFApplication.getInstance());
    }

    /**
     * Used by screenshots test
     */
    @SuppressWarnings("unused")
    public static Context setLocale(Locale locale) {
        return setLocale(OFFApplication.getInstance(), locale);
    }

    public static String getLanguage(Context context) {
        String lang = getLanguageInPreferences(context, Locale.getDefault().getLanguage());
        if (lang.contains("-")) {
            lang = lang.split("-")[0];
        }
        return lang;
    }

    public static Locale getLocale(Context context) {
        Resources resources = context.getResources();
        Configuration configuration = resources.getConfiguration();
        final Locale locale = configuration.locale;
        return locale == null ? Locale.getDefault() : locale;
    }

    public static Context setLocale(Context context, String language) {
        PreferenceManager.getDefaultSharedPreferences(context)
            .edit()
            .putString(SELECTED_LANGUAGE, language)
            .apply();

        Locale locale = getLocale(language);
        return setLocale(context, locale);
    }

    private static Context setLocale(Context context, Locale locale) {
        if (locale == null) {
            return context;
        }
        PreferenceManager.getDefaultSharedPreferences(context)
            .edit()
            .putString(SELECTED_LANGUAGE, locale.getLanguage())
            .apply();

        Locale.setDefault(locale);

        Resources resources = context.getResources();

        Configuration configuration = resources.getConfiguration();
        if (Build.VERSION.SDK_INT >= 17) {
            configuration.setLocale(locale);
            context = context.createConfigurationContext(configuration);
        } else {
            configuration.locale = locale;
        }
        resources.updateConfiguration(configuration, resources.getDisplayMetrics());
        return context;
    }

    /**
     * Extract language and region from the locale string
     *
     * @param locale language
     * @return Locale from locale string
     */
    public static Locale getLocale(String locale) {
        String[] localeParts = locale.split("-");
        String language = localeParts[0];
        String country = localeParts.length == 2 ? localeParts[1] : "";
        Locale localeObj = null;
        if (locale.contains("+")) {
            localeParts = locale.split("\\+");
            language = localeParts[1];
            String script = localeParts[2];
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
                for (Locale checkLocale : Locale.getAvailableLocales()) {
                    if (checkLocale.getISO3Language().equals(language) && checkLocale.getCountry().equals(country) && checkLocale.getVariant().equals("")) {
                        localeObj = checkLocale;
                    }
                }
            } else {
                localeObj = new Locale.Builder().setLanguage(language).setRegion(country).setScript(script).build();
            }
        } else {
            localeObj = new Locale(language, country);
        }
        return localeObj;
    }

    private static String getLanguageInPreferences(Context context, String defaultLanguage) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        return preferences.getString(SELECTED_LANGUAGE, defaultLanguage);
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/utils/LocaleHelper.kt
/*
 * Copyright 2016-2020 Open Food Facts
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package openfoodfacts.github.scrachx.openfood.utils

import android.content.Context
import android.os.Build
import androidx.core.content.edit
import androidx.core.os.ConfigurationCompat
import androidx.preference.PreferenceManager
import openfoodfacts.github.scrachx.openfood.R
import openfoodfacts.github.scrachx.openfood.app.OFFApplication
import openfoodfacts.github.scrachx.openfood.network.ApiFields
import java.util.*

/**
 * This class is used to change your application locale and persist this change for the next time
 * that your app is going to be used.
 *
 *
 * You can also change the locale of your application on the fly by using the setLocale method.
 *
 *
 * Created by gunhansancar on 07/10/15.
 */
object LocaleHelper {
    private val SELECTED_LANGUAGE by lazy { OFFApplication.instance.getString(R.string.pref_language_key) }

    fun List<LanguageData>.find(language: String) = indexOfFirst { language == it.code }

    fun onCreate(context: Context, defaultLanguage: String = Locale.getDefault().language) =
            setLanguageInPrefs(context, getLanguageInPreferences(context, defaultLanguage))

    fun getLanguageData(codes: Collection<String?>, supported: Boolean) =
            codes.map { getLanguageData(it, supported) }.sorted().toMutableList()

    fun getLanguageData(code: String?, supported: Boolean) =
            getLocale(code).let { LanguageData(it.language, it.getDisplayName(it).capitalize(Locale.ROOT), supported) }

    /**
     * Used by screenshots test
     */
    internal fun setLanguageInPrefs(locale: Locale) = setContextLanguage(OFFApplication.instance, locale)

    fun getLanguage(context: Context?): String {
        var lang = getLanguageInPreferences(context, Locale.getDefault().language)
        if (lang.contains("-")) {
            lang = lang.split("-")[0]
        }
        return lang
    }

    fun getLCOrDefault(languageCode: String?) =
            if (!languageCode.isNullOrEmpty()) languageCode else ApiFields.Defaults.DEFAULT_LANGUAGE

    fun getLocaleFromContext(context: Context? = OFFApplication.instance): Locale {
        var locale: Locale? = null
        if (context != null) {
            val resources = context.resources
            if (resources != null) {
                val configuration = resources.configuration
                if (configuration != null) {
                    locale = ConfigurationCompat.getLocales(configuration)[0]
                }
            }
        }
        return locale ?: Locale.getDefault()
    }

    /**
     * Used by screenshots generator.
     *
     * @param context
     * @param locale
     */
    fun setContextLanguage(context: Context, locale: Locale): Context {
        var newContext = context
        PreferenceManager.getDefaultSharedPreferences(newContext).edit {
            putString(SELECTED_LANGUAGE, locale.language)
        }
        Locale.setDefault(locale)
        val resources = newContext.resources
        val configuration = resources.configuration
        if (Build.VERSION.SDK_INT >= 17) {
            configuration.setLocale(locale)
            newContext = newContext.createConfigurationContext(configuration)
        } else {
            configuration.locale = locale
        }
        resources.updateConfiguration(configuration, resources.displayMetrics)
        return newContext
    }

    fun setLanguageInPrefs(context: Context, language: String?): Context {
        PreferenceManager.getDefaultSharedPreferences(context).edit {
            putString(SELECTED_LANGUAGE, language)
            apply()
        }
        return setContextLanguage(context, getLocale(language))
    }

    /**
     * Extract language and region from the locale string
     *
     * @param locale language
     * @return Locale from locale string
     */
    fun getLocale(locale: String?): Locale {
        if (locale == null) return Locale.getDefault()

        var localeParts = locale.split("-").toTypedArray()
        var language = localeParts[0]
        val country = if (localeParts.size == 2) localeParts[1] else ""
        if (!locale.contains("+")) return Locale(language, country)

        localeParts = locale.split("+").toTypedArray()
        language = localeParts[1]
        val script = localeParts[2]
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            Locale.Builder().setLanguage(language).setRegion(country).setScript(script).build()
        } else {
            Locale.getAvailableLocales().firstOrNull {
                it.isO3Language == language && it.country == country && it.variant == ""
            } ?: Locale.getDefault()
        }
    }

    private fun getLanguageInPreferences(context: Context?, defaultLanguage: String) = context?.let {
        PreferenceManager.getDefaultSharedPreferences(it).getString(SELECTED_LANGUAGE, defaultLanguage)
    } ?: defaultLanguage

    class LanguageData internal constructor(
            val code: String,
            val name: String,
            val isSupported: Boolean
    ) : Comparable<LanguageData> {
        override fun toString() = "$name [$code]"

        override fun equals(other: Any?) = other is LanguageData && this.code == other.code

        override fun hashCode() = code.hashCode()

        override fun compareTo(other: LanguageData) = name.compareTo(other.name)
    }
}